### ISSUE TITLE
Add accessors to MongoDB\Collection options

### ISSUE THREAD

**Issue by FloatingMaster** [mongodb/mongo-php-library#375]
It looks a lot more comfortable to set options via setter then passing $options array to constructor:
```
$collection = new Collection($manager, $databaseName, $collectionName);
$collection->setWriteConcern(new WriteConcern('0, 0, FALSE'));
$collection->setTypeMap(['root'->self::getDocumentNamespace()]);
```
vs
`$collection = new Collection($manager, $databaseName, $collectionName, ['writeConcern' => new WriteConcern('0, 0, FALSE')]);`

And in some cases it's useful to modify some Collection's options at runtime, for example, typeMap (we are using in our own serialization mechanizm).

**Comment by FloatingMaster** [comment]
#353 Add WriteConcern getter
#235 Extensive use of private methods instead of protected

**Comment by FloatingMaster** [comment]
#235

**Comment by bjori** [comment]
I don't this its a good idea to add such setters.

If the object becomes mutable like this you can never be sure of its current state and will have to reset it to the state required by individual function in case any other function mutated the object.

If you'd like to be able to modify the object in a specific function I'd rather recommend adding a way to get a new cloned object, with one modification, that can be disposed of after use.

```
$collection = new Collection($manager, $databaseName, $collectionName);
$collectionWithDifferentWriteConcernAndTypeMap = $collection->withWriteConcern(...)->withTypeMap([...]);
```

**Comment by lindelius** [comment]
Also, remember that you can specify those options for every single operation. So, if you need to "change it at runtime", do it for those reads/writes that requires different options.

**Comment by FloatingMaster** [comment]
Ok, I'm agreed that mutate object isn't a good idea and we can use some kind of Factory or Builder class to build Collection instance with needed options.

**Comment by petr-buchin** [comment]
I agree that `Collection` class does not need to have setters for the options, but at least it should have getters. 
Here is a real use case:
suppose I want to wrap a `Collection` instance to log each executed query/command and display this information in some kind of debug toolbar. I can write a class for this, for example:

```php
class LoggableCollection
{
    private $collection;

    public function __construct(Collection $collection)
    {
        $this->collection = $collection;
    }

    public function find($filter, array $options = [])
    {
        $log = new Query($filter, $options);
        $cursor = parent::find($filter, $options);
        $log->setEndTime(microtime(true) * 1000);
        $this->addLog($log);

        return $cursor;
    }

    // ...
}
```

Suppose I use this class as follows:

```
$mongoCollection = (new \MongoDB\Client)->selectCollection('test', 'test');
$collection = new LoggableCollection($mongoCollection);
$collection->find(['foo' => 'bar']);
```

As you see - I don't pass any options to `LoggableCollection::find()`. This means that this query will be logged without options. But under the hood `\MongoDB\Collection` will add it's default options to the query. And since `\MongoDB\Collection` does not have getters for read concern, write concern and read preference, there is no way to know, what options will be used in queries/commands. Off course `__debugInfo()` method can be used to get this options, but this is a dirty hack.

**Comment by jmikola** [comment]
@petr-buchin: Would [PHPLIB-271](https://jira.mongodb.org/browse/PHPLIB-271) be sufficient?

**Comment by petr-buchin** [comment]
@jmikola totally!
Can you tell when approximately this will be done?
I see that this task has status "Needs triage". 
Does it mean that this feature is not approved yet? 
If it is approved, though, I would be happy to create a merge request to have this feature asap.

**Comment by jmikola** [comment]
@petr-buchin: Feel free to create a PR for this. "Needs triage" is the default state for new issues, but I would like to get this into the upcoming 1.2.0 release.

If you create PR for the code changes, it would also be much appreciated if you could take a crack at creating the new documentation pages, since that is actually the bulk of the work here. If not, no worries -- I can always get to that after merging the code changes :)

**Comment by petr-buchin** [comment]
@jmikola , I created a PR, but it needs review, since documentation pages may be not ok :)

**Comment by jmikola** [comment]
Closing this, as option accessors were implemented in https://github.com/mongodb/mongo-php-library/pull/401 (released in 1.2.0).
